csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedExpr.qll

```diff
@@ -0,0 +1,2432 @@
+        opcode instanceof Opcode::Constant and
+        resultType = getConstantType()
+      ) or
+      (
+        tag = CrementOpTag() and
+        opcode = getOpcode() and
+        resultType = getResultType()
+      ) or
+      (
+        tag = CrementStoreTag() and
+        opcode instanceof Opcode::Store and
+        resultType = getResultType()
+      )
+    )
+  }
+
+  override final Instruction getInstructionOperand(InstructionTag tag,
+      OperandTag operandTag) {
+    (
+      tag = CrementLoadTag() and
+      (
+        (
+          operandTag instanceof AddressOperandTag and
+          result = getOperand().getResult()
+        ) or
+        (
+          operandTag instanceof LoadOperandTag and
+          result = getEnclosingFunction().getUnmodeledDefinitionInstruction()
+        )
+      )
+    ) or
+    (
+      tag = CrementOpTag() and
+      (
+        (
+          operandTag instanceof LeftOperandTag and
+          result = getInstruction(CrementLoadTag())
+        ) or
+        (
+          operandTag instanceof RightOperandTag and
+          result = getInstruction(CrementConstantTag())
+        )
+      )
+    ) or
+    (
+      tag = CrementStoreTag() and
+      (
+        (
+          operandTag instanceof AddressOperandTag and
+          result = getOperand().getResult()
+        ) or
+        (
+          operandTag instanceof StoreValueOperandTag and
+          result = getInstruction(CrementOpTag())
+        )
+      )
+    )
+  }
+
+  override final Instruction getFirstInstruction() {
+    result = getOperand().getFirstInstruction()
+  }
+
+  override final Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    kind instanceof GotoEdge and
+    (
+      (
+        tag = CrementLoadTag() and
+        result = getInstruction(CrementConstantTag())
+      ) or
+      (
+        tag = CrementConstantTag() and
+        result = getInstruction(CrementOpTag())
+      ) or
+      (
+        tag = CrementOpTag() and
+        result = getInstruction(CrementStoreTag())
+      ) or
+      (
+        tag = CrementStoreTag() and
+        result = getParent().getChildSuccessor(this)
+      )
+    )
+  }
+
+  override final Instruction getChildSuccessor(TranslatedElement child) {
+    child = getOperand() and result = getInstruction(CrementLoadTag())
+  }
+
+  override final int getInstructionElementSize(InstructionTag tag) {
+    tag = CrementOpTag() and
+    (
+      getOpcode() instanceof Opcode::PointerAdd or
+      getOpcode() instanceof Opcode::PointerSub
+    ) and
+    result = 4 //max(getResultType().(PointerType).getSize())
+  }
+
+  final TranslatedExpr getOperand() {
+    result = getTranslatedExpr(expr.getOperand())
+  }
+
+  final Opcode getOpcode() {
+    exists(Type resultType |
+      resultType = getResultType() and
+      (
+        (
+          expr instanceof IncrementOperation and
+          if resultType instanceof PointerType then
+            result instanceof Opcode::PointerAdd
```

Comment:
I kept the `PointerAdd` since it is a bit different from `Add` since sometimes it accepts an immediate operand, namely the size of the element it points to, and I thought that might be useful in the future.

Reply:
? @dave-bartolomeo